sdk/appconfiguration/app-configuration/types/app-configuration.d.ts

```diff
@@ -0,0 +1,387 @@
+    /**
+     * Lists revisions of a set of keys, optionally filtered by key names,
+     * labels and accept datetime.
+     *
+     * Example code:
+     * ```ts
+     * const revisionsIterator = await client.listRevisions({ keys: ["MyKey"] });
+     * ```
+     * @param options Optional parameters for the request.
+     */
+    listRevisions(options?: ListRevisionsOptions): PagedAsyncIterableIterator<ConfigurationSetting, ListRevisionsPage>;
+    private getListRevisionsIterator;
+    private listRevisionsByPage;
+    /**
+     * Sets the value of a key in the Azure App Configuration service, allowing for an optional etag.
+     * @param key The name of the key.
+     * @param configurationSetting A configuration value.
+     * @param options Optional parameters for the request.
+     *
+     * Example code:
+     * ```ts
+     * await client.setConfigurationSetting({ key: "MyKey", value: "MyValue" });
+     * ```
+     */
+    setConfigurationSetting(configurationSetting: SetConfigurationSettingParam, options?: SetConfigurationSettingOptions): Promise<SetConfigurationSettingResponse>;
+    /**
+     * Sets a key's value to read only
+     * @param id The id of the configuration setting to set to read-only.
+     */
+    setReadOnly(id: ConfigurationSettingId, options?: SetReadOnlyOptions): Promise<SetReadOnlyResponse>;
+    /**
+     * Makes the key's value writable again
+     * @param id The id of the configuration setting to make writable.
+     */
+    clearReadOnly(id: ConfigurationSettingId, options?: ClearReadOnlyOptions): Promise<ClearReadOnlyResponse>;
+}
+
+/**
+ * Options for clearReadOnly
+ */
+export declare interface ClearReadOnlyOptions extends HttpOnlyIfUnchangedField, RequestOptionsBase {
+}
+
+/**
+ * Response when clearing the read-only status from a value
+ */
+export declare interface ClearReadOnlyResponse extends ConfigurationSetting, SyncTokenHeaderField, HttpResponseField<SyncTokenHeaderField> {
+}
+
+/**
+ * Configuration setting with extra metadata from the server, indicating
+ * its etag, whether it is currently readOnly and when it was last modified.
+ */
+export declare interface ConfigurationSetting extends ConfigurationSettingParam {
+    /**
+     * Whether or not the setting is read-only
+     */
+    readOnly: boolean;
+    /**
+     * The date when this setting was last modified
+     */
+    lastModified?: Date;
+}
+
+/**
+ * Fields that uniquely identify a configuration setting
+ */
+export declare interface ConfigurationSettingId {
+    /**
+     * The key for this setting
+     */
+    key: string;
+    /**
+     * The label for this setting. Leaving this undefined means this
+     * setting does not have a label.
+     */
+    label?: string;
+    /**
+     * The etag for this setting
+     */
+    etag?: string;
+}
+
+/**
+ * Necessary fields for updating or creating a new configuration setting
+ */
+export declare interface ConfigurationSettingParam extends ConfigurationSettingId {
```

Comment:
I wonder if this is one of those cases where everything is optional, but really you should specify at least one. There should be a TS way to express this...

Reply:
'key' is required - I'm suffering a bit in this class because it's the one generically used across all the APIs and some of them (listConfigurationSettings, for example) allow you to select which fields are actually filled in. 

I could see us trying to make a separation there - @chradek  mentioned the same thing so it could be worth it and I don't think it'd be a horrible amount of work (just some runtime copying, which I'm already having to do to support locked/readOnly translation).

Reply:
Leaving as is.